### PR TITLE
Unified mixed precision API + backend performance improvements

### PR DESCRIPTION
**Pull request by mcarilli**
PRing for comparison.  Single-GPU and distributed tests pass for the full cross product of supported pure and mixed precision options.  Any code using the old API should also continue to work properly for now, although it will be discouraged and eventually removed.

Documentation still needs some work.

### CODE REVIEW

File: apex/fp16_utils/__init__.py [hunk 1]
```diff
@@ -14,3 +14,5 @@
 
 from .fp16_optimizer import FP16_Optimizer
 from .loss_scaler import LossScaler, DynamicLossScaler
+
+test = 1
```
Comment (mcarilli) [inline apex/fp16_utils/__init__.py]: delete

**Comment by dave-epstein** [comment]
Once this is merged, is there a recommended process for porting over old code that uses FP16_Optimizer?

**Comment by mcarilli** [comment]
@dave-epstein of course!  Working on that now.  It‘ll be really easy.

**Comment by dave-epstein** [comment]
> @dave-epstein of course! Working on that now. It‘ll be really easy.

Great! Thanks so much, this is a fantastic library :)

**Comment by mcarilli** [comment]
@dave-epstein I'm still working on making the documentation more comprehensive but in the meantime you can check out the reworked Imagenet example https://github.com/NVIDIA/apex/tree/api_refactor/examples/imagenet.

`--opt-level O2` is equivalent to FP16_Optimizer with dynamic loss scaling. (In fact, it calls directly into FP16_Optimizer under the hood.  Once I get everyone everyone on the new API I can stop maintaining FP16_Optimizer as a separate entity.)

**Pull request merged**